rust/sealevel/programs/mailbox/src/processor.rs

```diff
@@ -0,0 +1,855 @@
+            b"-",
+            &init.local_domain.to_le_bytes(),
+            b"-",
+            b"authority",
+            &[auth_bump],
+        ]],
+    )?;
+
+    let inbox_account = next_account_info(accounts_iter)?;
+    let (inbox_key, inbox_bump) = Pubkey::find_program_address(
+        &[
+            b"hyperlane",
+            b"-",
+            &init.local_domain.to_le_bytes(),
+            b"-",
+            b"inbox",
+        ],
+        program_id,
+    );
+    if &inbox_key != inbox_account.key {
+        return Err(ProgramError::InvalidArgument);
+    }
+    invoke_signed(
+        &system_instruction::create_account(
+            payer_account.key,
+            inbox_account.key,
+            Rent::default().minimum_balance(mailbox_size.try_into().unwrap()),
+            mailbox_size,
+            program_id,
+        ),
+        &[
+            system_program.clone(),
+            payer_account.clone(),
+            inbox_account.clone(),
+        ],
+        &[&[
+            b"hyperlane",
+            b"-",
+            &init.local_domain.to_le_bytes(),
+            b"-",
+            b"inbox",
+            &[inbox_bump],
+        ]],
+    )?;
+
+    let outbox_account = next_account_info(accounts_iter)?;
+    let (outbox_key, outbox_bump) = Pubkey::find_program_address(
+        &[
+            b"hyperlane",
+            b"-",
+            &init.local_domain.to_le_bytes(),
+            b"-",
+            b"outbox",
+        ],
+        program_id,
+    );
+    if &outbox_key != outbox_account.key {
+        return Err(ProgramError::InvalidArgument);
+    }
+    invoke_signed(
+        &system_instruction::create_account(
+            payer_account.key,
+            outbox_account.key,
+            Rent::default().minimum_balance(mailbox_size.try_into().unwrap()),
+            mailbox_size,
+            program_id,
+        ),
+        &[
+            system_program.clone(),
+            payer_account.clone(),
+            outbox_account.clone(),
+        ],
+        &[&[
+            b"hyperlane",
+            b"-",
+            &init.local_domain.to_le_bytes(),
+            b"-",
+            b"outbox",
+            &[outbox_bump],
+        ]],
+    )?;
+
+    let inbox = Inbox {
+        local_domain: init.local_domain,
+        auth_bump_seed: auth_bump,
+        inbox_bump_seed: inbox_bump,
+        ..Default::default()
+    };
+    InboxAccount::from(inbox).store(inbox_account, true)?;
+
+    let outbox = Outbox {
+        local_domain: init.local_domain,
+        auth_bump_seed: auth_bump,
+        outbox_bump_seed: outbox_bump,
+        ..Default::default()
+    };
+    OutboxAccount::from(outbox).store(outbox_account, true)?;
+
+    Ok(())
+}
+
+// TODO add more strict checks on permissions for all accounts that are passed in, e.g., bail if
+// we expected an account to be read only but it is writable. Could build this into the AccountData
+// struct impl and provide more fetch methods.
+fn inbox_process(
```

Comment:
ah so the recipient cannot be the mailbox contract? We could add one however there is a finite compute budget and if the payer sends a txn that will cause the contract to recurse infinitely, it will burn through the compute budget. It may make more sense to leave as is to disincentivize making recursive calls like that.